KVM Dirty log ring interface

davidkleymann opened the pull request:
### Summary of the PR

Add support for dirty page tracking via the Dirty ring interface. Adds KvmDirtyLogRing structure for keeping track of the indices and the base pointer to the shared memory buffer. Implements iterating over dirty pages, thereby harvesting them. Implements reset_dirty_rings on VmFd to trigger recycling of dirty ring buffer elements by the kernel after processing. Adds the dirty_log_ring field to VcpuFd.

This is a draft that needs some review and improvements, I'm hereby asking for suggestions for improving the following remaining weaknesses:

1. If any vCPU is running while harvesting dirty pages, the user of KvmDirtyLogRing needs to call reset_dirty_rings *before* reading contents of dirty pages. This is currently the users responsibility, which allows for calling reset_dirty_rings after all dirty pages have been read by the VMM for cases where all vCPUs are halted, but this may be a confusing interface for other cases.

More info on the interface:
https://www.kernel.org/doc/html/latest/virt/kvm/api.html#kvm-cap-dirty-log-ring-kvm-cap-dirty-log-ring-acq-rel

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR have Signed-Off-By trailers (with
  `git commit -s`), and the commit message has max 60 characters for the
  summary and max 75 characters for each description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

roypat commented on kvm-ioctls/src/ioctls/vcpu.rs:
````diff
@@ -2104,6 +2106,65 @@ impl VcpuFd {
+    ///     }
+    /// }
+    /// ```
+    pub fn dirty_log_ring_iter(&mut self) -> Option<&mut KvmDirtyLogRing> {
````
we can hide the actual iterator type here by just returning `impl Iterator<Item = blablabla>`. Then maybe the entire struct doesnt need to be exported?

roypat commented on kvm-ioctls/src/ioctls/vcpu.rs:
```diff
@@ -2106,8 +2106,7 @@ impl VcpuFd {
 
-    /// Maps the coalesced MMIO ring page. This allows reading entries from
-    /// the ring via [`coalesced_mmio_read()`](VcpuFd::coalesced_mmio_read).
+    /// Maps the KVM dirty log ring. 
```
this should go into the commit that introduced the function

roypat commented on kvm-ioctls/src/ioctls/mod.rs:
and all the changes to this file should go into the commit that introduced struct KvmDirtyLogRing

roypat reviewed (changes requested):
please squash your commits so that we do not have later commits be fixups for previous commits.

roypat commented:
Also, do we need to do anything about KVM_CAP_DIRTY_LOG_RING_WITH_BITMAP?

roypat commented:
Also sorry for the late response, we were all preparing for / traveling to KVM Forum last week!

roypat commented on kvm-ioctls/src/ioctls/mod.rs:
```diff
@@ -29,6 +29,78 @@ pub mod vm;
+}
+
+impl Iterator for KvmDirtyLogRing {
+    type Item = (u32, u64);
+    fn next(&mut self) -> Option<Self::Item> {
+        let i = self.next_dirty & self.mask;
+        unsafe {
+            let gfn = self.gfns.add(i as usize).as_mut();
+            if (*gfn).flags & KVM_DIRTY_GFN_F_DIRTY == 0 {
+                // next_dirty stays the same, it will become the next dirty element
+                return None;
+            } else {
+                self.next_dirty += 1;
+                (*gfn).flags ^= KVM_DIRTY_GFN_F_RESET;
+                return Some(((*gfn).slot, (*gfn).offset));
+            }
+        }
+    }
```
I think even if we don't support arm64 initially (although, is it really that much more complicated to also do the KVM_CAP_DIRTY_LOG_RING_WITH_BITMAP thing? Sorry, it's been a while since I looked at all this), I think we should make the code forward compatible for when we _do_ add arm64 support.

davidkleymann committed:
kvm-ioctls: add dirty log ring support for all architectures

Implement dirty log ring interface with `enable_dirty_log_ring` and
`dirty_log_ring_iter` methods. Enable `VmFd` `enable_cap` and ioctl imports
on all architectures. Add memory fences in iterator for proper
synchronization on weak memory consistency architectures.

Signed-off-by: David Kleymann <david@asymmetric.re>

davidkleymann committed:
kvm-ioctls: add dirty log ring support for all architectures

Implement dirty log ring interface with `enable_dirty_log_ring` and
`dirty_log_ring_iter` methods. Enable `VmFd` `enable_cap` and ioctl imports
on all architectures. Add memory fences in iterator for proper
synchronization on weak memory consistency architectures.

Signed-off-by: David Kleymann <david@asymmetric.re>

roypat commented:
> Why is enable_cap not available in aarch64? `#[cfg(any(target_arch = "x86_64", target_arch = "s390x", target_arch = "powerpc"))]`

according to the docs, it doesn't exist on arm: https://docs.kernel.org/virt/kvm/api.html#kvm-enable-cap . But it seems the docs are wrong there, because there is a generic implementation in the kernel for some capabilities (including the dirty log related ones) which is available on all arches. So we can just expose enable_cap() unconditionally i think.

davidkleymann commented:
> > Send and Sync are probably not safe on KvmDirtyLogRing because accesses are not stateless, due to the state we need to keep track of in user space (next_dirty)
> 
> As for this, `next_dirty` doesnt imply that it cannot be Send/Sync, because the traits dont mean that multiple threads can access the struct at the same time (thats always unsound in Rust's memory model), just that you can transfer ownership between threads, and that read-only references can be shared between threads. A problem would only arrive if we can also `Clone` this structure, because then the iterator can cause data races inside the mmap'd area from different Rust threads. But as long as for each vcpu we only allow creating a single instance of the iterator (using safe code that is), there are no problems.
> 
> Now, whether Send and Sync will actually be useful is a different matter, because I'm assuming you want them to be able to harvest the dirty ring while the vcpu is still running, but with the current API in this PR, you can only get a &mut reference to the ring buffer structure, which you cannot send to another thread anyway. So in this case, you'd need some API that lets you take ownership of the KvmDirtyRingLog structure (maybe store the one that gets created at vcpu creation time in an option, and then have a function that just wraps .take() on that option, keeping in mind that we must never allow safe code to get two owned versions of it for the same vcpu).

I needed Send and Sync to move the Vcpus to separate threads, but I only implemented what the rust compiler suggested and am still looking for a better solution for this unrelated problem.

zolutal commented:
Hi @davidkleymann , thanks for your work on this feature! I was testing it out and noticed that no VcpuExit is currently defined for `KVM_EXIT_DIRTY_RING_FULL`. Currently you just get a `VcpuExit::Unsupported(31)` when the ring buffer fills up. If it doesn't make it in this PR I can make one myself to add it, but I figured I'd mention it here.

roypat commented:
> Hi @davidkleymann , thanks for your work on this feature! I was testing it out and noticed that no VcpuExit is currently defined for `KVM_EXIT_DIRTY_RING_FULL`. Currently you just get a `VcpuExit::Unsupported(31)` when the ring buffer fills up. If it doesn't make it in this PR I can make one myself to add it, but I figured I'd mention it here.

that is a good catch, thanks! This should definitely be added in this PR :)

davidkleymann commented:
Thanks, rebased.

ShadowCurse commented on kvm-ioctls/src/ioctls/vm.rs:
````diff
@@ -2050,6 +2078,131 @@ impl VmFd {
+    /// // Create one vCPU with the ID=0 to cause a dirty log ring to be mapped.
+    /// let vcpu = vm.create_vcpu(0);
+    /// ```
+    pub fn enable_dirty_log_ring(&mut self, bytes: Option<i32>) -> Result<bool> {
````
How about instead of providing `bytes` as an argument, we provide a number of entries in the log we want? So `entries: Option<NonZeroU32>`? This will make it harder to make a mistake + less validation checks inside.
This can also be done for the `KvmDirtyLogRing::mmap_from_fd`

davidkleymann replied:
I can change it to providing number of entries but in that case it would also be helpful to implement something that wraps `vm.check_extension_raw(KVM_CAP_DIRTY_LOG_RING.into())` and returns the max number of entries.

ShadowCurse commented on .vscode/settings.json:
```diff
@@ -0,0 +1,44 @@
+{
```
pretty sure this file should not be here

davidkleymann commented:
When will this be merged?

roypat commented:
I've just trigged dependabot, which should give us a PR for fixing the CI failure, and then we can merge this immediately after! Sorry for the long wait :(

davidkleymann commented:
Thanks. No worries, just making sure it gets merged eventually :)

ludfjig reviewed (approved):
lgtm (the latest main branch merge commit should maybe have been a rebase instead?)

davidkleymann commented on kvm-ioctls/src/ioctls/vcpu.rs:
```diff
@@ -197,6 +199,9 @@ pub struct VcpuFd {
     /// A pointer to the coalesced MMIO page
     coalesced_mmio_ring: Option<KvmCoalescedIoRing>,
+    /// A pointer to the dirty log ring
+    #[allow(unused)]
```
fixed

davidkleymann commented:
Rebased and removed an unneccessary #[allow(unused)].
